[TensorExpr] Implement LoopNest::computeAt

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#36112 [TensorExpr] Implement LoopNest::computeAt**
* #35120 [TensorExpr] Bounds inference



Differential Revision: [D20885662](https://our.internmc.facebook.com/intern/diff/D20885662)

Commit:
[WIP][TensorExpr] Implement LoopNest::computeAt

[ghstack-poisoned]

Commit:
Update on "[WIP][TensorExpr] Implement LoopNest::computeAt"




[ghstack-poisoned]

Commit:
clang-tidy on "[WIP][TensorExpr] Implement LoopNest::computeAt"




[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Implement LoopNest::computeAt"




Differential Revision: [D20885662](https://our.internmc.facebook.com/intern/diff/D20885662)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Implement LoopNest::computeAt"




Differential Revision: [D20885662](https://our.internmc.facebook.com/intern/diff/D20885662)

[ghstack-poisoned]

zheng-xq commented on test/cpp/tensorexpr/test_loopnest.cpp:
```diff
@@ -765,6 +766,159 @@ void testBoundsInference_5() {
+  std::ostringstream oss;
+  oss << *s;
+
+  const std::string& verification_pattern =
```
I am not a big fan of this string check pattern. It is difficult to strike a balance that checks enough, but not too much.

I would prefer manually construct the reference loop-nest, convert to string, and match two strings are equal.

ZolotukhinM replied:
I actually much prefer this than byte-to-byte comparison os strings, which is way too fragile and will break any time we change something internally. Using patterns I can specify what exactly to check for: e.g. a compute at needs to insert `Allocate` of a given size: I can check for that and at the same time ignore, say, the temp variable name which can change at any point. Another example that can easily change in future: for temporary computations we currently replicate the entire original loopnest even if iteration over some of the axis has a range of 1. E.g. we can generate:
```
for y in 0..1:
  for x in 0..W:
     temp[y,x] = ...
```
At some point we will start optimizing out loops `for y in 0..1` - this would also break the test if I check for the exact match of strings.

zheng-xq commented on torch/csrc/jit/tensorexpr/loopnest.cpp:
```diff
@@ -915,6 +917,275 @@ Stmt* FlattenIndexes(Stmt* s) {
+      st->mask());
+
+  // Construct the loop nest for the temp computation
+  for (size_t i = 0; i < dims.size(); i++) {
```
Add a TODO to test this against Nick's new reduction axis.

zheng-xq commented on test/cpp/tensorexpr/test_loopnest.cpp:
```diff
@@ -765,6 +766,159 @@ void testBoundsInference_5() {
+
+    assertAllEqual(c_data, c_ref);
+  }
+}
```
Add a TODO to test against cases where there are reduction axis.

zheng-xq commented on torch/csrc/jit/tensorexpr/loopnest.cpp:
```diff
@@ -915,6 +917,275 @@ Stmt* FlattenIndexes(Stmt* s) {
+  }
+
+  // Generate alloc/free stmts for 'temp'
+  Stmt* al = new Allocate(temp_buf->base_handle(), st->value()->dtype(), dims);
```
I am still thinking whether the temp should be here, or near the global scope. Maybe a TODO for us to revisit this later?

Commit:
Update on "[TensorExpr] Implement LoopNest::computeAt"




Differential Revision: [D20885662](https://our.internmc.facebook.com/intern/diff/D20885662)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Implement LoopNest::computeAt"




Differential Revision: [D20885662](https://our.internmc.facebook.com/intern/diff/D20885662)

[ghstack-poisoned]

Commit:
rebase on "[TensorExpr] Implement LoopNest::computeAt"




Differential Revision: [D20885662](https://our.internmc.facebook.com/intern/diff/D20885662)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Implement LoopNest::computeAt"




Differential Revision: [D20885662](https://our.internmc.facebook.com/intern/diff/D20885662)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Implement LoopNest::computeAt"




Differential Revision: [D20885662](https://our.internmc.facebook.com/intern/diff/D20885662)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Implement LoopNest::computeAt"




Differential Revision: [D20885662](https://our.internmc.facebook.com/intern/diff/D20885662)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Implement LoopNest::computeAt"




Differential Revision: [D20885662](https://our.internmc.facebook.com/intern/diff/D20885662)

[ghstack-poisoned]
